### PR TITLE
refactor(connector): add amount conversion framework to Shift4

### LINKED ISSUE #6122: [REFACTOR]: [SHIFT4] Add amount conversion framework to Shift4
### :memo: Feature Description

Currently, amounts are represented as `i64` values throughout the application. We want to introduce a `Unit` struct that explicitly states the denomination. A new type, `MinorUnit`, has been added to standardize the flow of amounts across the application. This type will now be used by all the connector flows.

Rather than handling conversions in each connector, we will centralize the conversion logic in one place within the core of the application.

### :hammer: Possible Implementation

- For each connector, we need to create an amount conversion function. Connectors will specify the format they require, and the core framework will handle the conversion accordingly.
- Connectors should invoke the `convert` function to receive the amount in their required format.
- Refer to the [connector documentation](https://dev.shift4.com/docs/api#cards) to determine the required amount format for each connector.
- You can refer [this PR](https://github.com/juspay/hyperswitch/pull/4825) for more context.

🔖 Note: All the changes needed should be contained within `hyperswitch/crates/router/src/connector/` , `crates/router/src/types/api.rs` , `crates/router/tests/connectors/`

### :package: Have you spent some time checking if this feature request has been raised before?

- [ ] I checked and didn't find a similar issue

### :package: Have you read the Contributing Guidelines?

- [ ] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### :package: Submission Process:
- Ask the maintainers for assignment of the issue, you can request for assignment by commenting on the issue itself.
- Once assigned, submit a pull request (PR).
- Maintainers will review and provide feedback, if any. 
- Maintainers can unassign issues due to inactivity, [read more here](https://github.com/juspay/hyperswitch/wiki/Hacktoberfest-Contribution-Rules).

Refer [here](https://github.com/juspay/hyperswitch/blob/main/docs/TERMS_OF_CONTEST.md) for Terms and conditions for the contest.

### PR DESCRIPTION
**Pull request by AnishKacham**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [X] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
This PR adds amount conversion framework to Shift4, for sending to connector.
Resolves #6122 

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Review by Sakilmostak — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/shift4.rs [hunk 1]
```diff
@@ -27,8 +31,18 @@ use crate::{
...
-pub struct Shift4;
+#[derive(Clone)]
+pub struct Shift4 {
+    amount_converter: &'static (dyn AmountConverter<Output = MinorUnit> + Sync),
+}
+
+impl Shift4 {
+    pub fn new() -> &'static Self {
+        &Self {
+            amount_converter: &StringMinorUnitForConnector,
```
[the remark this answers is not in the document]
Comment (AnishKacham) [reply crates/router/src/connector/shift4.rs]: corrected

File: crates/router/src/connector/shift4/transformers.rs [hunk 2]
```diff
@@ -205,7 +205,7 @@ where
         item: &types::RouterData<T, Req, types::PaymentsResponseData>,
     ) -> Result<Self, Self::Error> {
         let submit_for_settlement = item.request.is_automatic_capture()?;
-        let amount = item.request.get_amount_required()?.to_string();
+        let amount = MinorUnit::new(item.request.get_amount_required()?);
```
[the remark this answers is not in the document]
Comment (AnishKacham) [reply crates/router/src/connector/shift4/transformers.rs]: done

File: crates/router/src/connector/shift4/transformers.rs [hunk 3]
```diff
@@ -447,7 +447,7 @@ impl<T> TryFrom<&types::RouterData<T, types::CompleteAuthorizeData, types::Payme
                 let card_token: Shift4CardToken =
                     to_connector_meta(item.request.connector_meta.clone())?;
                 Ok(Self {
-                    amount: item.request.amount.to_string(),
+                    amount: MinorUnit::new(item.request.amount),
```
[the remark this answers is not in the document]
Comment (AnishKacham) [reply crates/router/src/connector/shift4/transformers.rs]: done

File: crates/router/src/connector/shift4/transformers.rs [hunk 4]
```diff
@@ -648,7 +648,7 @@ pub struct Shift4WebhookObjectResource {
 pub struct Shift4NonThreeDsResponse {
     pub id: String,
     pub currency: String,
-    pub amount: u32,
+    pub amount: MinorUnit,
```
[the remark this answers is not in the document]
Comment (AnishKacham) [reply crates/router/src/connector/shift4/transformers.rs]: removed the usage

File: crates/router/src/connector/shift4/transformers.rs [hunk 5]
```diff
@@ -811,15 +811,15 @@ impl<T, F>
...
+    amount: MinorUnit,
 }
 
 impl<F> TryFrom<&types::RefundsRouterData<F>> for Shift4RefundRequest {
     type Error = Error;
     fn try_from(item: &types::RefundsRouterData<F>) -> Result<Self, Self::Error> {
         Ok(Self {
             charge_id: item.request.connector_transaction_id.clone(),
-            amount: item.request.refund_amount,
+            amount: MinorUnit::new(item.request.refund_amount),
```
[the remark this answers is not in the document]
Comment (AnishKacham) [reply crates/router/src/connector/shift4/transformers.rs]: done

File: crates/router/src/connector/shift4/transformers.rs [hunk 6]
```diff
@@ -837,7 +837,7 @@ impl From<Shift4RefundStatus> for enums::RefundStatus {
 #[derive(Default, Debug, Clone, Serialize, Deserialize)]
 pub struct RefundResponse {
     pub id: String,
-    pub amount: i64,
+    pub amount: MinorUnit,
```
[the remark this answers is not in the document]
Comment (AnishKacham) [reply crates/router/src/connector/shift4/transformers.rs]: removed the usage

**Review by Sakilmostak — DISMISSED** [review]
LGTM!

**Comment by deepanshu-iiitu** [comment]
@AnishKacham 
Please resolve build and clippy errors. Use `cargo build` and `cargo clippy`

**Comment by AnishKacham** [comment]
Fixed the clippy and build errors, only thing remaining is a dead-code warning because of not using a function that was previously being used, should that be addressed as well?  

@Sakilmostak @deepanshu-iiitu

**Comment by swangi-kumari** [comment]
@AnishKacham ,
Can you please resolve the failing checks.
Run `just clippy` and resolve all errors and warnings.

**Review by swangi-kumari — APPROVED** [review]

**Review by Sakilmostak — APPROVED** [review]

File: crates/router/src/connector/shift4/transformers.rs [hunk 7]
```diff
@@ -95,7 +101,7 @@ impl Shift4AuthorizePreprocessingCommon for types::PaymentsPreProcessingData {
 }
 #[derive(Debug, Serialize)]
 pub struct Shift4PaymentsRequest {
-    amount: String,
+    amount: MinorUnit,
```
Comment (Narayanbhat166) [inline crates/router/src/connector/shift4/transformers.rs:104]: @swangi-kumari @deepanshu-iiitu should this be `StringMinorUnit`? can you please confirm this?

Comment (deepanshu-iiitu) [reply crates/router/src/connector/shift4/transformers.rs:104]: Actually it needs to be integer only. I think our previous integration was not correct.
https://dev.shift4.com/docs/api

**Review by deepanshu-iiitu — APPROVED** [review]

**Review by dracarys18 — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Pull request merged**